### PR TITLE
import: check if file is accessible

### LINKED ISSUE #8598: darktable wont quit if .jpg|.xmp  are unopenable ( no premission rwx=0000 )
place some pics in a directory
chmod 0000 *
run darktable (3.4.1)
import the directory
get errors about not being able to open files for reading ( no permissions )
close darktable
( now just wait, and wait, and wait, and ...... ) 

___________________________________________________________________
fedora linux 33
darktable build with src rpm.

Other things go wrong before you try/give up to close up.
like 
[New LWP 555353]

(darktable:555157): GLib-GObject-WARNING **: 17:24:11.713: invalid unclassed pointer in cast to 'GtkDarktableExpander'

** (darktable:555157): CRITICAL **: 17:24:11.713: dtgtk_expander_get_expanded: assertion 'DTGTK_IS_EXPANDER(expander)' failed
[LWP 555353 exited]

### PR DESCRIPTION
**Pull request by HansBull**
Should fix #8598

### CODE REVIEW

**Review by jenshannoschwalm — COMMENTED** [review]
Just two things i want to mention

1. At other places we always use the GLib wrappers like g_fopen ...
2. I think this shot is too fast. At least this doesn't work for windows systems, i have no personal experience about windows at all but there are many reports (i remember one about writing sidecars) just about this problem. Checking for: is a file readable/writable. I think we would need a windows specific implementation and always use a wrapper doing the os specific function. If you look at  https://developer.gnome.org/glib/stable/glib-File-Utilities.html ...

**Comment by HansBull** [comment]
Oh, indeed, there is really a `g_access`.
What about all the other existing uses of `access` in lut3d.c line 1561, database.c line 3450, 3457, image.c line 1265 ... etc? Don't they work on Windows systems?

**Comment by jenshannoschwalm** [comment]
I can not test and have no idea about how to implement a proper `win_access` function. But i think it would be good to have all `access` functions identified and replaced by a `dt_file_access` (or whatever it may be called) wrapper function hoping for a windows dev to step in to implement it correctly.

**Comment by TurboGit** [comment]
> I can not test and have no idea about how to implement a proper `win_access` function. But i think it would be good to have all `access` functions identified and replaced by a `dt_file_access` (or whatever it may be called) wrapper function hoping for a windows dev to step in to implement it correctly.

Won't g_acces() will work ok on Windows?

**Comment by jenshannoschwalm** [comment]
We had issues about the file access on sdcards on windows or on other devices. I also had a short look at the issue for the pr and found the docs about all the g_functions. The problem about file permission checking is exactly described there. It also gives a reference to do such things in a windows-proper way.

**Comment by jenshannoschwalm** [comment]
pinging @peterbud what is your position on this? (I think you did most of the win-specific code so far.)

**Comment by jenshannoschwalm** [comment]
Wouldn't we have to use the "access tokens" ?

**Comment by cryptomilk** [comment]
Take a look at the code a few lines above:

```
  if(!normalized_filename
     || !g_file_test(normalized_filename, G_FILE_TEST_IS_REGULAR)
     || dt_util_get_file_size(normalized_filename) == 0)
  {
```

With this patch, we are doing **three** stat() calls where one would be enough! Just call g_lstat() and check for regular file, file size and readability please.

**Comment by jenshannoschwalm** [comment]
Hi @HansBull i think we should/can ignore any windows specific stuff. I did some more reading and probably g_access should be fine. Are you going to "finish" this pr to get it read-to-merge?

File: src/common/image.c [hunk 1]
```diff
@@ -1394,7 +1394,8 @@ static uint32_t _image_import_internal(const int32_t film_id, const char *filena
   char *normalized_filename = dt_util_normalize_path(filename);
   if(!normalized_filename
      || !g_file_test(normalized_filename, G_FILE_TEST_IS_REGULAR)
-     || dt_util_get_file_size(normalized_filename) == 0)
+     || dt_util_get_file_size(normalized_filename) == 0
```
Comment (jenshannoschwalm) [inline src/common/image.c:1397]: If you look into the definition of dt_util_get_file_size in utility.c, i think you should compare to <1 instead of ==0

Comment (cryptomilk) [reply src/common/image.c:1397]: This is 3 times calling stat() on the file path. Please reduce it to one call!

**Comment by jenshannoschwalm** [comment]
@TurboGit could you mark this as a bug / issue that must be resolved for 3.6 (just to not forget it)

**Comment by TurboGit** [comment]
@jenshannoschwalm : Done.

File: src/common/image.c (hunk 1 above)
Comment (jenshannoschwalm) [reply src/common/image.c:1397]: @cryptomilk you mentioned this twice. file size and regular file can be done with one stat, sure. But access()? isn't there more involved than just bit-mask checking for uid/gid?

**Comment by jenshannoschwalm** [comment]
Hi Hans, what about a somewhat better testing for a valid image or xmp file like
```
gboolean dt_util_test_image_file(const char *filename)
{
  if(g_access(filename, R_OK)) return FALSE;
#ifdef _WIN32
  struct _stati64 stats;
  if(_stati64(filename, &stats)) return FALSE;
#else
  struct stat stats;
  if(stat(filename, &stats)) return FALSE;
#endif

  const gboolean regular = (S_ISREG(stats.st_mode)) != 0;
  const gboolean size_ok = stats.st_size > 0;
  return regular & size_ok;
}
```
This seems - for me - to be a safe way to test for a valid file (readable, regular and likely having content), it could be used in `_image_import_internal` and also while checking for duplicates in
- `dt_image_find_duplicates` to check for valid xmp
- `generate_thumbnail_cache`to check for existing thumbs

Unfortunately i don't know about a way how to do the g_access equivalent from the stat struct in a portable way as @cryptomilk suggested.

**Pull request closed without merging**